tests/tests_pytorch/loops/optimization/test_closure.py

```diff
@@ -43,3 +48,103 @@ def step(self, closure=None):
     trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
     with pytest.raises(MisconfigurationException, match="The closure hasn't been executed"):
         trainer.fit(model)
+
+
+def test_closure_with_no_grad_optimizer(tmpdir):
```

Comment:
Could you simplify the optimizer definition so that the issue still reproduces?
Something like this?
```py
class NoGradAdamW(torch.optim.AdamW):
    @torch.no_grad()
    def step(self, *args, **kwargs):
        return super().step(*args, **kwargs)

class TestModel(BoringModel):
    def training_step(self, batch, batch_idx):
        assert torch.is_grad_enabled()
        return super().training_step(batch, batch_idx)

    def configure_optimizers(self):
        return NoGradAdamW(self.layer.parameters(), lr=0.1)
            
trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
model = TestModel()
trainer.fit(model)

```

Reply:
fixed, please check.